src/System.Net.Primitives/src/System/Net/IPv6AddressHelper.cs

```diff
@@ -0,0 +1,472 @@
+            for (int i = fromInclusive; i < toExclusive; i++)
+            {
+                if (numbers[i] == 0)
+                {
+                    currentSequenceLength++;
+                    if (currentSequenceLength > longestSequenceLength)
+                    {
+                        longestSequenceLength = currentSequenceLength;
+                        longestSequenceStart = i - currentSequenceLength + 1;
+                    }
+                }
+                else
+                {
+                    currentSequenceLength = 0;
+                }
+            }
+
+            return longestSequenceLength > 1 ?
+                (longestSequenceStart, longestSequenceStart + longestSequenceLength) :
+                (-1, -1);
+        }
+
+        // Returns true if the IPv6 address should be formatted with an embedded IPv4 address:
+        // ::192.168.1.1
+        internal unsafe static bool ShouldHaveIpv4Embedded(ushort[] numbers)
+        {
+            // 0:0 : 0:0 : x:x : x.x.x.x
+            if (numbers[0] == 0 && numbers[1] == 0 && numbers[2] == 0 && numbers[3] == 0 && numbers[6] != 0)
+            {
+                // RFC 5952 Section 5 - 0:0 : 0:0 : 0:[0 | FFFF] : x.x.x.x
+                if (numbers[4] == 0 && (numbers[5] == 0 || numbers[5] == 0xFFFF))
+                {
+                    return true;
+                }
+                // SIIT - 0:0 : 0:0 : FFFF:0 : x.x.x.x
+                else if (numbers[4] == 0xFFFF && numbers[5] == 0)
+                {
+                    return true;
+                }
+            }
+
+            // ISATAP
+            return numbers[4] == 0 && numbers[5] == 0x5EFE;
+        }
+
+        //
+        // IsValidStrict
+        //
+        //  Determine whether a name is a valid IPv6 address. Rules are:
+        //
+        //   *  8 groups of 16-bit hex numbers, separated by ':'
+        //   *  a *single* run of zeros can be compressed using the symbol '::'
+        //   *  an optional string of a ScopeID delimited by '%'
+        //   *  the last 32 bits in an address can be represented as an IPv4 address
+        //
+        //  Difference between IsValid() and IsValidStrict() is that IsValid() expects part of the string to 
+        //  be ipv6 address where as IsValidStrict() expects strict ipv6 address.
+        //
+        // Inputs:
+        //  <argument>  name
+        //      IPv6 address in string format
+        //
+        // Outputs:
+        //  Nothing
+        //
+        // Assumes:
+        //  the correct name is terminated by  ']' character
+        //
+        // Returns:
+        //  true if <name> is IPv6  address, else false
+        //
+        // Throws:
+        //  Nothing
+        //
+
+        //  Remarks: MUST NOT be used unless all input indexes are verified and trusted.
+        //           start must be next to '[' position, or error is reported
+        internal unsafe static bool IsValidStrict(char* name, int start, ref int end)
```

Comment:
Yeah, it's both brittle and inefficient.  But it's what desktop has, and I think it's best to stick with it for now rather than rewriting the whole parsing routine.